third_party/java/arcs/build_defs/internal/schemas.bzl

```diff
@@ -164,16 +159,110 @@ def arcs_kt_schema(
         )
     return {"outs": outs, "deps": outdeps}
 
+def _arcs_ts_preproc_impl(ctx):
+    outputs = [
+        ctx.actions.declare_file(x.basename, sibling = x)
+        for x in ctx.files.srcs
+    ]
+
+    ctx.actions.run_shell(
+        inputs = ctx.files.srcs,
+        outputs = outputs,
+        arguments = [src.path for src in ctx.files.srcs],
+        command = """
+        for i in "$@"; do
+          out="$(pwd)/bazel-out/host/bin/$i"
+          sed -e 's/-web.js/-node.js/g' $i > "$out"
+        done
+        """,
+    )
+
+    return DefaultInfo(files = depset(outputs))
+
+arcs_ts_preprocessing = rule(
+    implementation = _arcs_ts_preproc_impl,
+    attrs = {
+        "srcs": attr.label_list(allow_files = [".ts", ".d.ts"]),
+    },
+    doc = """Converts to node platforms for the arcs ts runtime.""",
+)
+
+def _schema2pkg_impl(ctx):
+    output_name = ctx.label.name + "_GeneratedSchemas.%s.kt" % ctx.attr.platform
+    out = ctx.actions.declare_file(output_name)
+
+    args = ctx.actions.args()
+
+    if ctx.attr.platform == "cpp":
+        args.add("--cpp")
+        if ctx.attr.platform != "wasm":
+            fail("'wasm' platform is only supported with 'kotlin' language.")
+        if ctx.attr.test_harness:
+            fail("'test_harness' is only supported with 'kotlin' language.")
+    else:
+        args.add("--kotlin")
+
+    if ctx.attr.platform == "wasm":
+        args.add("--wasm")
+
+    if ctx.attr.test_harness:
+        args.add("--test_harness")
+
+    args.add_all("--outdir", [out.dirname])
+    args.add_all("--outfile", [output_name])
+    args.add_all([src.path for src in ctx.files.srcs])
+
+    ctx.actions.run(
+        inputs = ctx.files.srcs,
+        outputs = [out],
+        arguments = [args],
+        tools = ctx.files.data,
+        executable = ctx.executable._compiler,
+    )
+
+    return [DefaultInfo(files = depset([out]))]
+
+schema2pkg = rule(
+    implementation = _schema2pkg_impl,
+    attrs = {
+        "srcs": attr.label_list(allow_files = [".arcs"]),
+        "deps": attr.label_list(),
+        "data": attr.label_list(allow_files = True),
+        "platform": attr.string(
+            values = ["jvm", "wasm"],
+            default = "jvm",
+        ),
+        "lang": attr.string(
+            values = ["kotlin", "cpp"],
+            default = "kotlin",
+        ),
+        "test_harness": attr.bool(
+            default = False,
+            doc = """whether to output a particle test harness only (requires lang=kotlin)""",
+        ),
+        "_compiler": attr.label(
+            cfg = "host",
+            default = Label("//src/tools:schema2pkg"),
+            allow_files = True,
+            executable = True,
+        ),
+    },
+    doc = """Stand-alone schema2* tool.""",
```

Comment:
The fact that it is standalone is only relevant for this PR. Once you land this PR, there will be no other way than stand-alone, so it's not really relevant. When writing docs you should think in terms of what makes sense to convey to the reader who will be reading this code tomorrow and a year from now. Typically you should not convey the history of what code went through, unless it is important in understanding what the code does - I don't think it is the case here.

Given this is now *the* rule for schema2pkg, you should write a more proper documentation for it, at least explain the arguments, usage etc.